### ISSUE TITLE
Suspected memory leak in Sprout

### ISSUE THREAD

**Issue by GAB-MS** [Metaswitch/sprout#1493]
#### Symptoms
While operating a Sprout EC2 node and running Perf tests on the cluster containing this node, a suspected memory leak was observed in the sprout process. This Sprout node is using mostly default configuration as per manual Clearwater install instructions, on a t2.small node, but with BGCF functionality disabled as it is routing through an external BGCF. The setup is not configured to use any application servers.

The Perf test for this system loops 1,000 callers (that register, send an invite, update, and a bye) through the network. Over about a 2 hour observation period, the test performed about 20,000 calls - 20 iterations per caller - and during this period, memory usage by the sprout process rose steadily from 28.6% to 50.7% as reported by top. (This period of observation started after the stress tests had been running for around half an hour, so some memory had already leaked at this point.) CPU usage on the Sprout node was around 6% for the duration of the stress testing. 

#### Impact
The memory leak will eventually cause Sprout to crash. 

#### Release and environment 
Running Sprout package version 1.0-160613.162259.

**Comment by mirw** [comment]
I've attached the output from massif.
[massif.out.20186.txt](https://github.com/Metaswitch/sprout/files/378657/massif.out.20186.txt)

**Comment by huws** [comment]
On 2016/08/03 I ran the Clearwater SIP stress tests against a standard minimal deployment.  I ran this for 185,000 calls, and sprout showed no signs of any leak.  This test does not route any traffic off-network, though.  Hopefully I'll soon be able to re-run with off-network traffic, which should give us a clue whether this is triggered by the presence of an external BGCF, or the off-network traffic itself.

**Comment by mirw** [comment]
Any update on this?  Did you manage to repro?

**Comment by huws** [comment]
No, sorry - our plans changed, so I ended up not doing this.  I can try this relatively easily though - I'll give it a go tonight.

**Comment by huws** [comment]
I hit problems running these tests last night.  I'll try to have another go today.
FTR - I was able to run regular Clearwater stress against a system with an external BGCF without seeing any evidence of a memory leak in sprout (after ~50,000 calls).  That's not a very interesting result though, as the regular Clearwater stress tests don't include any off-network traffic.

**Comment by huws** [comment]
I've not been able to run the off-net tests on a Clearwater system with the internal BGCF.  These tests have not been kept in a working state, and I've not been able to get them running again.  Sorting this out is going to require more time than I can absorb.  Is this issue of any value without a working repro scenario?

**Comment by mirw** [comment]
OK, so we have an issue where we believe the defect still exists, but we no longer have the repro setup.  I think that means we should lower the priority and the next action is for us to try to set up our own repro - if we can't repro, we should send back to Huw.

**Comment by eleanor-merry** [comment]
I've updated the live test docs for the offnet tests (https://github.com/Metaswitch/clearwater-live-test/pull/138), but I've not had a chance to run the tests through to detect a memory leak. Unclaiming as I'm unlikely to work on this soon,
